src/smc-webapp/project/history/log.tsx

```diff
@@ -0,0 +1,346 @@
+//##############################################################################
+//
+//    CoCalc: Collaborative Calculation in the Cloud
+//
+//    Copyright (C) 2016, Sagemath Inc.
+//
+//    This program is free software: you can redistribute it and/or modify
+//    it under the terms of the GNU General Public License as published by
+//    the Free Software Foundation, either version 3 of the License, or
+//    (at your option) any later version.
+//
+//    This program is distributed in the hope that it will be useful,
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//    GNU General Public License for more details.
+//
+//    You should have received a copy of the GNU General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+//##############################################################################
+
+import * as misc from "smc-util/misc";
+
+import * as immutable from "immutable";
+
+import {
+  React,
+  rtypes,
+  rclass,
+  redux,
+  TypedMap,
+  Rendered
+} from "../../app-framework";
+
+import { Button } from "react-bootstrap";
+
+import { Icon, Loading } from "../../r_misc";
+import { WindowedList } from "../../r_misc/windowed-list";
+
+import { LogSearch } from "./search";
+import { LogEntry } from "./log-entry";
+import { ProjectLogMap, EventRecord } from "./types";
+import { ProjectActions } from "smc-webapp/project_store";
+import { UserMap } from "smc-webapp/todo-types";
+
+interface ReactProps {
+  project_id: string;
+  name: string;
+  search?: string;
+  actions: ProjectActions;
+}
+
+interface ReduxProps {
+  project_log?: ProjectLogMap;
+  project_log_all?: ProjectLogMap;
+  search?: string;
+
+  user_map?: UserMap;
+  get_name: (user_id: string) => string;
+}
+
+interface State {
+  cursor_index: number;
+}
+
+export const ProjectLog = rclass<ReactProps>(
+  class ProjectLog extends React.Component<ReactProps & ReduxProps, State> {
+    static reduxProps = ({ name }): Record<string, any> => {
+      return {
+        [name]: {
+          project_log: rtypes.immutable,
+          project_log_all: rtypes.immutable,
+          search: rtypes.string
+        },
+        users: {
+          user_map: rtypes.immutable,
+          get_name: rtypes.func
+        }
+      };
+    };
+
+    static defaultProps = { search: "" }; // search that user has requested
+
+    private windowed_list_ref: React.RefObject<any>;
+    private _log: immutable.List<TypedMap<EventRecord>>;
```

Comment:
FYI, in all such cases I'm happy with deleting the leading underscores in private typescript variables.  We don't need to use that (python-ish) convention, since we have typescript. :-)